core/src/apps/cardano/helpers/lazy_cbor_collection.py

```diff
@@ -0,0 +1,114 @@
+import ustruct as struct
+from micropython import const
+
+from apps.common import cbor
+
+if False:
+    from typing import Generic, Iterator, TypeVar
+
+    T = TypeVar("T")
+else:
+    T = 0  # type: ignore
+    Generic = [object]  # type: ignore
+
+
+_CBOR_TYPE_ARRAY = const(0b100 << 5)
+_CBOR_TYPE_MAP = const(0b101 << 5)
+
+
+class PauseIteration:
+    """
+    Used to signal to consumers of LazyCborCollection that all available data has been consumed for now.
+    """
+
+    pass
+
+
+class LazyCborCollection(Generic[T]):
+    """
+    A base class for LazyCborCollections. Contains basic functionality for a lazy collection. Enables iteration over
+    items as they are being added (one by one) thus never storing them all at once.
+    """
+
+    class EmptyElement:
+        """
+        Serves as a replacement for None since None can be a valid CBOR item.
+        """
+
+        pass
+
+    def __init__(self, size: int) -> None:
+        self.size = size
+        self.counter = 0
+        self.element: T | LazyCborCollection.EmptyElement = self.EmptyElement()
+
+    def __len__(self) -> int:
+        return self.size
+
+    def is_filled(self) -> bool:
+        return self.counter == self.size
+
+    def __iter__(self) -> Iterator[T]:
+        return self
+
+    def __next__(self) -> T:
+        if self.is_filled() and isinstance(self.element, self.EmptyElement):
+            raise StopIteration
+        assert not isinstance(
+            self.element, self.EmptyElement
+        ), "The next element hasn't been added yet"
+        element, self.element = self.element, self.EmptyElement()
+        return element
+
+    def append_item(self, item: T) -> None:
+        assert isinstance(
+            self.element, self.EmptyElement
+        ), "The previous element hasn't been read yet"
+        assert self.counter < self.size, "Item appended to a filled collection"
+        self.element = item
+        self.counter += 1
+
+    def cbor_serialize(self) -> Iterator[bytes] | PauseIteration:
+        raise NotImplementedError
+
+
+class LazyCborDict(LazyCborCollection):
```

Comment:
`append_item` should probably assert that the item being added is a pair.

Reply:
I've added a `_validate_item` function to `LazyCborCollection` which gets overridden in both `LazyCborDict` and `LazyCborList` although it does nothing for the list. I'm not sure it's the best approach though.